staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go

```diff
@@ -133,3 +136,164 @@ func TestWatchErrResultNotBlockAfterCancel(t *testing.T) {
 	cancel()
 	wg.Wait()
 }
+
+func TestWatchChanSync(t *testing.T) {
+	origCtx, store, _ := testSetup(t)
+	nativeKV := store.client.KV
+	initList, err := initStoreData(origCtx, store)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testCases := []struct {
+		name             string
+		watchKey         string
+		watcherMaxLimit  int64
+		addGetReactor    bool
+		expectEventCount int
+		expectGetCount   int
+	}{
+		{
+			name:             "None of the current objects match watchKey: sync with empty page",
+			watchKey:         "/pods/test/",
+			watcherMaxLimit:  1,
+			addGetReactor:    false,
+			expectEventCount: 0,
+			expectGetCount:   1,
+		},
+		{
+			name:             "The number of current objects is less than defaultWatcherMaxLimit: sync with one page",
+			watchKey:         "/pods/",
+			watcherMaxLimit:  3,
+			addGetReactor:    false,
+			expectEventCount: 2,
+			expectGetCount:   1,
+		},
+		{
+			name:             "a new item added to etcd before returning a second page is not returned: sync with two page",
+			watchKey:         "/pods/",
+			watcherMaxLimit:  1,
+			addGetReactor:    true,
+			expectEventCount: 2,
+			expectGetCount:   2,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			defaultWatcherMaxLimit = testCase.watcherMaxLimit
+
+			kvWrapper := newEtcdClientKVWrapper(nativeKV)
+			if testCase.addGetReactor {
+				kvWrapper.getReactors = append(kvWrapper.getReactors, func() {
+					barThird := &example.Pod{ObjectMeta: metav1.ObjectMeta{Namespace: "third", Name: "bar"}}
+					podKey := fmt.Sprintf("/pods/%s/%s", barThird.Namespace, barThird.Name)
+					storedObj := &example.Pod{}
+
+					err := store.Create(context.Background(), podKey, barThird, storedObj, 0)
```

Comment:
if we can manipulate the page size (via `watcherMaxLimit`) maybe we could simplify and add the items during a test setup.

Reply:
No - because we want to add them in the middle of the request to validate that consistency works as expected.

Reply:
we could test both cases :)

Reply:
> if we can manipulate the page size (via `watcherMaxLimit`) maybe we could simplify and add the items during a test setup.

I think this case is already covered, since we init etcd data during a test setup